File: lua/vimade/state/win.lua
local M = {}

local bit = require('bit')
local BIT_BAND = bit.band
local BIT_BOR = bit.bor

local NAMESPACE = require('vimade.state.namespace')
local REAL_NAMESPACE = require('vimade.state.real_namespace')
local LINK = require('vimade.config_helpers.link')
local BLOCKLIST = require('vimade.config_helpers.blocklist')
local COLOR_UTIL = require('vimade.util.color')
local COMPAT = require('vimade.util.compat')
local TYPE = require('vimade.util.type')
local FOCUS = require('vimade.focus')

local HIGHLIGHTERS = {
  require('vimade.highlighters.terminal'),
  require('vimade.highlighters.namespace'),
}

local GLOBALS
local FADER

M.__init = function(args)
  FADER = args.FADER
  GLOBALS = args.GLOBALS
  FADER.on('tick:before', function ()
    M.current = nil
  end)
end

M.cache = {}

local _update_state = function (next, current, state)
  local modified = false
  for field, value in pairs(next) do
    if current[field] ~= value then
      current[field] = value
      modified = true
    end
  end
  return modified and state or GLOBALS.READY
end

-- return without updating state
M.get = function(winid)
  return M.cache[winid]
end

M.unhighlight = function (win)
  win.nc = false
  for _, highlighter in ipairs(HIGHLIGHTERS) do
    if not highlighter.is_type or highlighter.is_type(win) then
      highlighter.unhighlight(win)
    end
  end
end

M.cleanup = function (active_winids)
  for winid, value in pairs(M.cache) do
    if active_winids[winid] == nil then
      M.cache[winid] = nil
      NAMESPACE.clear_winid(winid)
    end
  end
end

M.__create = function (winid)
  if M.cache[winid] == nil then
    local area = FOCUS.get(winid)
    local win = {
      winid = winid,
      bufnr = nil,
      terminal = nil,
      width = -1,
      height = -1,
      linked = false,
      blocked = false,
      blocked_highlights = {},
      nc = false,
      hi_key = '',
      is_active_win = nil,
      is_active_buf = nil,
      area_owner = nil,
      area = area,
      is_focus = area and area.is_focus, -- should never change
      is_mark = area and area.is_mark, -- should never change
      style_state = {
        animations = {},
        custom = {}
      },
      timestamps = {
        active_win = 0,
        active_buf = 0,
        nc = 0,
      },
      winhl = nil,
      real_ns = nil,
      ns = nil,
      state = GLOBALS.READY,
      style = {},
      terminal = false,
      terminal_match = nil,
      -- raw global styles that were used to build this win
      _global_style = {},
      buf_opts = nil,
      buf_vars = nil,
      win_opts = nil,
      win_vars = nil,
    }
    M.cache[winid] = win
  end
  return M.cache[winid]
end

M.refresh_active = function (winid)
  local win = M.refresh(winid, true)
  return win
end

M.refresh = function (winid, is_active)
  if not vim.api.nvim_win_is_valid(winid) then
    return
  end
  local win = M.__create(winid)
  if is_active then
    M.current = win
  end

  win.state = GLOBALS.READY
  win.winid = winid
  win.bufnr = vim.api.nvim_win_get_buf(winid)
  win.buf_name = vim.api.nvim_buf_get_name(win.bufnr)
  win.win_type = vim.fn.win_gettype(win.winid)
  win.win_config = vim.api.nvim_win_get_config(win.winid)
  win.buf_opts = vim.bo[win.bufnr]
  win.buf_vars = vim.b[win.bufnr]
  win.win_opts = vim.wo[win.winid]
  win.win_vars = vim.w[win.winid]
  win.terminal = win.buf_opts.buftype == 'terminal'

  local is_active_win = win.winid == GLOBALS.current.winid
  local is_active_buf = win.bufnr == GLOBALS.current.bufnr

  -- area timestamps and 'active' status is inherited from the owner window
  -- it does not need to be set or relied on.
  if not win.area then
    if is_active_win ~= win.is_active_win then
      win.is_active_win = is_active_win
      win.timestamps.active_win = GLOBALS.now
    end
    if is_active_buf ~= win.is_active_buf then
      win.is_active_buf = is_active_buf
      win.timestamps.active_win = GLOBALS.now
    end
  else
    local area_owner = M.get(win.area.source.winid)
    if area_owner then
      win.area_owner = area_owner
    end
    if  win.area.hide then
      return
    end
  end

  local can_nc = not (vim.b[win.bufnr].vimade_disabled == 1
    or vim.w[win.winid].vimade_disabled == 1)

  local blocked = false
  -- areas should skip the blocklist in all scenarios
  if not win.area then
    for key, value in pairs(GLOBALS.blocklist) do
      if type(value) == 'table' then
        blocked = BLOCKLIST.DEFAULT(win, M.current, value)
      elseif type(value) == 'function' then
        blocked = value(win, M.current)
      end
      if blocked == true then
        can_nc = false
        break
      end
    end
    win.blocked = blocked
  end

  if blocked then
    return M.unhighlight(win)
  end

  local should_nc = false
  local hi_active = can_nc and GLOBALS.vimade_fade_active
  local hi_inactive = can_nc

  if GLOBALS.nc_windows and win.is_active_win then
    should_nc = hi_active
  elseif GLOBALS.nc_buffers and win.is_active_buf then
    should_nc = hi_active
  elseif (GLOBALS.nc_windows or GLOBALS.nc_buffers) then
    should_nc = hi_inactive
  end

  local linked = false
  if M.current and not is_active and not blocked then
    for key, value in pairs(GLOBALS.link) do
      if type(value) == 'table' then
        linked = LINK.DEFAULT(win, M.current, value)
      elseif type(value) == 'function' then
        linked = value(win, M.current)
      end
      if linked == true then
        should_nc = hi_active
        break
      end
    end
  end
  win.linked = linked

  -- check namespaces
  -- TODO this should actually occur before blocked
  local active_ns = COMPAT.nvim_get_hl_ns({winid = win.winid})
  local winhl
  local real_ns
  if NAMESPACE.is_vimade_ns(active_ns) == true then
    real_ns = win.real_ns
    winhl = win.winhl
  else
    winhl = vim.wo[win.winid].winhl
    real_ns = winhl and 0 or (active_ns == -1 and 0 or active_ns)
  end
  win.real_ns = real_ns
  win.winhl = winhl
  vim.w[win.winid]._vimade_real_ns = real_ns

  if (should_nc and not win.nc) or (not should_nc and win.nc) then
    win.timestamps.nc = GLOBALS.now
  end

  win.state = BIT_BOR(win.state, _update_state({
    nc = should_nc
  }, win, GLOBALS.CHANGED))

  local basebg = GLOBALS.basebg
  local basebg_key = ''
  if type(basebg) == 'function' then
    basebg = basebg(win, M.current)
  end
  if basebg then
    basebg = COLOR_UTIL.to24b(basebg)
  end
  if type(basebg) == 'number' then
    win.basebg = basebg
    basebg_key = 'bg(' .. basebg .. ')'
  else
    win.basebg = nil
  end


  local rerun_style = false
  if #win._global_style ~= #GLOBALS.style then
    rerun_style = true
  else
    for i, s in ipairs(GLOBALS.style) do
      if win._global_style[i] ~= s then
        rerun_style = true
        break
      end
    end
  end
  if rerun_style == true then
    win.style = {}
    win._global_style = {}
    for i, s in ipairs(GLOBALS.style) do
      win._global_style[i] = s
      win.style[i] = s.attach(win, win.style_state)
    end
  end

  local hi_key = (winhl or win.real_ns) .. ':' 
  hi_key = hi_key .. ':' .. basebg_key

  -- this is to separate out the logic for inactive vs active.  we can use this for highlighting
  -- the active ns in the future
  local style_active = 0
  if not blocked then
    for i, s in ipairs(win.style) do
      s.before(win, win.style_state)
      local style_key = s.key(win, style_state)
      if string.len(style_key) > 0 then
        if style_active > 0 then
          hi_key = hi_key .. '#'
        end
        hi_key = hi_key .. style_key
        style_active = style_active + 1
      end
    end
    win.blocked_highlights = BLOCKLIST.HIGHLIGHTS(win, M.current)
    hi_key = hi_key .. ':bh(' .. BLOCKLIST.TO_HIGHLIGHTS_KEY(win.blocked_highlights) .. ')'
  end

  local redraw = false
  if style_active > 0
    and (not win.ns
    or REAL_NAMESPACE.is_desync(win.ns.real)
    or not GLOBALS.nohlcheck
    or win.hi_key ~= hi_key
    or BIT_BAND(GLOBALS.tick_state, GLOBALS.RECALCULATE) > 0) then
    local ns = NAMESPACE.get_replacement(win, real_ns, hi_key, false)
    if ns.modified then
      redraw = true
    end
    if ns.modified or ns ~= win.ns or win.hi_key ~= hi_key then
      win.state = BIT_BOR(GLOBALS.CHANGED, win.state)
    end
    win.ns = ns
  end

  win.hi_key = hi_key

  if win.ns and win.ns.real.complete_highlights then
    for _, highlighter in ipairs(HIGHLIGHTERS) do
      if not highlighter.is_type or highlighter.is_type(win) then
        if style_active == 0 then
          highlighter.unhighlight(win)
        else
          highlighter.highlight(win, redraw)
        end
      end
    end
  end
  return win
end

return M

File: lib/vimade/state/win.py
import sys
import time
M = sys.modules[__name__]

import vim

from vimade import highlighter as HIGHLIGHTER
from vimade.state import globals as GLOBALS
from vimade.state import namespace as NAMESPACE
from vimade.config_helpers import link as LINK
from vimade.config_helpers import blocklist as BLOCKLIST
from vimade.util import color as COLOR_UTIL
from vimade.util import ipc as IPC
from vimade.util.promise import Promise,all

IS_NVIM = GLOBALS.is_nvim
FADER = None
M.cache = {}
M.current = None
M.diffs = 0
M.links = 0

def __init(args):
  global FADER
  FADER = args['FADER']
  def before():
    M.current = None
    M.links = 0
    M.diffs = 0
  FADER.on('tick:before', before)
M.__init = __init

class WinDeps(Promise):
  def __init__(self, win, is_active):
    Promise.__init__(self)
    self.win = win
    self.wincolor = None
    self.winhl = ''
    self.is_active = is_active
    self.get()

  def _apply_win_color(self):
    win = self.win
    if IS_NVIM:
      def set_winhl(value):
        self.winhl = value
        if len(value):
          for group in value.split(','):
            (source, target) = group.split(':')
            if source == 'Normal' and win.is_active_win:
              self.wincolor = target
              break
            elif source == 'NormalNC' and not win.is_active_win:
              self.wincolor = target
              break
        if self.wincolor == None:
          self.wincolor = 'Normal' if win.is_active_win else 'NormalNC'
      return IPC.batch_eval_and_return('gettabwinvar(%d,%d,"&winhl")'%(win.tabnr, win.winid)).then(set_winhl)
    else:
      def set_wincolor(value):
        self.wincolor = value
      return IPC.batch_eval_and_return('gettabwinvar(%d,%d,"&wincolor")'%(win.tabnr, win.winnr)).then(set_wincolor)
  def _apply_remaining_config(self):
    def next(config):
      config.append(self.wincolor)
      self.resolve(config)

    win = self.win
    winid = win.winid
    tabnr = win.tabnr
    winnr = win.winnr
    bufnr = win.bufnr

    all([
        IPC.batch_eval_and_return('['+
        ','.join([
          'gettabwinvar(%d,%d,"&wrap")' % (tabnr, winnr),
          'gettabwinvar(%d,%d,"&buftype")' % (tabnr, winnr),
          'gettabwinvar(%d,%d,"vimade_disabled")' % (tabnr, winnr),
          'getbufvar(%d, "vimade_disabled")' % bufnr,
          'getbufvar(%d, "&diff")' % bufnr,
          'gettabwinvar(%d,%d,"current_syntax")' % (tabnr, winnr),
          'gettabwinvar(%d,%d,"&syntax")' % (tabnr, winnr),
          'gettabwinvar(%d,%d,"&tabstop")' % (tabnr, winnr),
          'gettabwinvar(%d,%d,"&conceallevel")' % (tabnr, winnr),
          'nvim_win_get_config('+str(winid)+')' if HAS_NVIM_WIN_GET_CONFIG else '{}',
          'win_gettype('+str(winid)+')' if HAS_WIN_GETTYPE else '""',
        ]) + ']'),
       Promise().resolve(self.winhl),
       HIGHLIGHTER.get_hl_for_ids(self.win, [self.wincolor or GLOBALS.normalid, GLOBALS.normalid])\
     ]).then(next)
  def get(self):
    def next(val):
      self._apply_remaining_config()
    result = self._apply_win_color()
    if type(result) == Promise:
      result.then(next)
    else:
      next(None)
    return self


# TODO move these into features
HAS_NVIM_WIN_GET_CONFIG = True if int(IPC.eval_and_return('exists("*nvim_win_get_config")')) else False
HAS_NVIM_GET_HL_NS = True if int(IPC.eval_and_return('exists("*nvim_get_hl_ns")')) else False
HAS_WIN_GETTYPE = True if int(vim.eval('exists("*win_gettype")')) else False

def _get_values(input):
  if type(input) == list:
    return list
  elif type(input) == dict:
    return input.values()

def _update_state(next, current, state):
  modified = False
  for field, value in next.items():
    if current[field] != value:
      current[field] = value
      modified = True
  return state if modified else GLOBALS.READY

def get(wininfo):
  return M.cache.get(int(wininfo['winid']))

def cleanup(wininfos):
  map = {}
  for wininfo in wininfos:
    map[int(wininfo['winid'])] = True
  cache_keys = [key for key in cache.keys()]
  for winid in cache_keys:
    if not winid in map:
      cache[winid].cleanup()

def unhighlight(winid):
  win = M.cache.get(winid)
  if win:
    # set to a value that forces the window to have changed state (non-bool)
    win.nc = None
    if win.ns:
      win.ns.unhighlight()

def refresh_active(wininfo):
  return refresh(wininfo, True)

def refresh(wininfo, is_active = False):
  winid = int(wininfo['winid'])
  win = M.cache.get(winid)
  if not win:
    win = M.cache[winid] = WinState(wininfo)
  if is_active:
    M.current = win
  return win.refresh(wininfo, is_active)

class WinState(object):
  def __init__(self, wininfo):
    self.__internal = {
      # common
      'winid': int(wininfo['winid']),
      'winnr': None,
      'bufnr': None,
      'tabnr': None,
      'buf_name': None,
      'win_config': None,
      'win_type': None,
      'buf_vars': None,
      'buf_opts': None,
      'win_vars': None,
      'win_opts': None,
      'win_config': None,
      'basebg': None,
      'linked': False,
      'blocked': False,
      'nc': False,
      'should_nc': False,
      'fadelevel': None,
      'matchpriority': -1,
      'tint': None,
      'is_active_win': None,
      'is_active_buf': None,
      'state': GLOBALS.READY,
      'style': [],
      'style_active': 0,
      'style_state': {
        'animations': {},
        'custom': {},
      },
      'timestamps': {
        'active_win': 0,
        'active_buf': 0,
        'nc': 0,
      },
      '_global_style': [],
      # python-only (needed to due to difference in renderes)
      'diff': None,
      'window': None,
      'buffer': None,
      'coords_key': None,
      'hi_key': None,
      'ns': None,
      'height': -1,
      'width': -1,
      'matches': [],
      'cursor': (-1, -1),
      'wrap': False,
      'conceallevel': 0,
      'tabstop': None,
      'buftype': None,
      'syntax': None,
      'size_changed': False,
      'cursor_changed': False,
      'is_explorer': False,
      'is_minimap': False,
      'topline': -1,
      'botline': -1,
      'textoff': -1,
      'wincolor': None,
      'wincolorhl': {'id':-1, 'name': '', 'ctermfg': None, 'ctermbg': None, 'fg': None, 'bg': None, 'sp': None},
      'winhl': '',
      'original_wincolor': '',
      'original_winhl': '',
      'vimade_winhl': None,
    }
  def __getattribute__(self, key):
    if key == '_WinState__internal':
      return object.__getattribute__(self, key)
    elif not key in self.__internal:
      return object.__getattribute__(self, key)
    else:
      return object.__getattribute__(self, '_WinState__internal')[key]
  def __setattr__(self, key, value):
    if key == '_WinState__internal':
      object.__setattr__(self, key, value)
    else:
      object.__getattribute__(self, '_WinState__internal')[key] = value
    return value
  def __getitem__(self, key):
    if key == '__internal':
      return self.__internal
    else:
      return self.__internal[key]
  def __setitem__(self, key, value):
    self.__internal[key] = value
  def cleanup(self):
    del M.cache[self.winid]
    if self.ns:
      self.ns.cleanup()
  def get_window(self):
    try:
      if self.window:
        return self.window
      if not self.window:
        for tabpage in vim.tabpages:
          if tabpage.number == self.tabnr:
            for window in tabpage.windows:
              if window.number == self.winnr:
                self.window = window
                return self.window
    except:
      return None
    return None

  def refresh(self, wininfo, is_active):
    self.state = GLOBALS.READY
    self.winid = winid = int(wininfo['winid'])
    self.winnr = winnr = int(wininfo['winnr'])
    self.tabnr = tabnr = int(wininfo['tabnr'])
    self.bufnr = bufnr = int(wininfo['bufnr'])

    is_active_win = GLOBALS.current['winid'] == self.winid
    is_active_buf = GLOBALS.current['bufnr'] == self.bufnr

    if is_active_win != self.is_active_win:
      self.is_active_win = is_active_win
      self.timestamps['active_win'] = GLOBALS.now
    if is_active_buf != self.is_active_buf:
      self.is_active_buf = is_active_buf
      self.timestamps['active_buf'] = GLOBALS.now

    window = self.get_window()
    if window == None:
      return

    self.buffer = window.buffer
    self.buf_name = self.buffer.name

    if not self.ns:
      self.ns = NAMESPACE.Namespace(self)
    self.ns.refresh()

    def next(config):
      self.apply_async_config(config, wininfo, is_active)
      promise.resolve(self)

    promise = Promise()
    WinDeps(self, is_active).then(next)
    return promise
     

  def apply_async_config(self, config, wininfo, is_active):
    ((wrap,
         buftype,
         win_disabled,
         buf_disabled,
         diff,
         win_syntax,
         buf_syntax,
         tabstop,
         conceallevel,
         win_config,
         win_type,
     ),
     winhl,
     (wincolorhl, normalhl),
     wincolor
     ) = config

    window = self.window
    height = window.height
    width = window.width
    cursor = window.cursor
    syntax = win_syntax if win_syntax else buf_syntax

    self.win_config = win_config
    self.win_type = win_type
    self.buf_vars = self.buffer.vars
    self.buf_opts = self.buffer.options
    self.win_vars = self.window.vars
    self.win_opts = self.window.options
    self.diff = bool(int(diff))
    self.wincolor = wincolor

    if self.diff:
      M.diffs += 1

    was_vimade_wincolor = False
    if wincolor.startswith('vimade_'):
      wincolor = self.original_wincolor or ('NormalNC' if (IS_NVIM and self.is_active_win) else 'Normal')
      # this can happen because vim passes the wincolor over to new splits, not sure why it does this
      # but requires special handling otherwise target bg color is unknown
      if not self.original_wincolor:
        self.original_wincolor = wincolor # (based on the condition above this would be a Normal type)
        wincolorhl = normalhl
      else:
        was_vimade_wincolor = True
    else:
      self.original_wincolor = wincolor or ('NormalNC' if (IS_NVIM and self.is_active_win) else 'Normal')

    if not 'vimade_' in winhl:
      if self.original_winhl != winhl:
        self.state |= _update_state({
          'original_winhl': winhl
        }, self, GLOBALS.CHANGED |  GLOBALS.INVALIDATE_HIGHLIGHTS)
        self.vimade_winhl = None
    else:
      self.vimade_winhl = winhl
    self.winhl = winhl

    is_explorer = 'coc-explorer' in self.buf_name or 'NERD' in self.buf_name
    is_minimap = 'vim-minimap' in self.buf_name or '-MINIMAP-' in self.buf_name

    wincolorhl = HIGHLIGHTER.defaultWincolorHi(wincolorhl, normalhl)

    if not was_vimade_wincolor:
      self.wincolorhl = wincolorhl
    else:
      wincolorhl = self.wincolorhl


    # requires additional potentially more fading
    self.state |= _update_state({
      'syntax': syntax,
      'size_changed': height != self.height or width != self.width,
      'cursor_changed': cursor[0] != self.cursor[0] or cursor[1] != self.cursor[1],
      'wrap': int(wrap),
      'tabstop': int(tabstop),
      'is_explorer': is_explorer,
      'is_minimap': is_minimap,
      'conceallevel': int(conceallevel),
      'topline': int(wininfo['topline']),
      'botline': int(wininfo['botline']),
      'textoff': int(wininfo['textoff']),
    }, self, GLOBALS.CHANGED)


    # changes that don't necessarily trigger an update
    self.height = window.height
    self.width = window.width
    self.cursor = window.cursor

    can_nc = not win_disabled and not buf_disabled
    blocked = False
    for value in _get_values(GLOBALS.blocklist):
      if type(value) == dict:
        blocked = BLOCKLIST.DEFAULT(self, current, value)
      elif callable(value):
        blocked = value(self, current)
      if blocked == True:
        can_nc = False
        break

    should_nc = False
    hi_active = can_nc and GLOBALS.vimade_fade_active
    hi_inactive = can_nc

    if GLOBALS.nc_windows and self.is_active_win:
      should_nc = hi_active
    elif GLOBALS.nc_buffers and self.is_active_buf:
      should_nc = hi_active
    else:
      should_nc = hi_inactive

    linked = False
    if current and not is_active:
      for value in _get_values(GLOBALS.link):
        if type(value) == dict:
          linked = LINK.DEFAULT(self, current, value)
        elif callable(value):
          linked = value(self, current)
        if linked == True:
          should_nc = hi_active
          M.links += 1
          break

    self.blocked = blocked
    self.linked = linked

    if (should_nc and not self.nc) or (not should_nc and self.nc):
      self.timestamps['nc'] = GLOBALS.now

    self.should_nc = should_nc

    # buffers manage the collected synID()'s so that we don't need to reprocess
    # these should only change if the syntax has changed (this can happen)
    # per window, using :help ownsyntax
    coords_key = syntax 

    self.state |= _update_state({
      'coords_key': coords_key,
      'syntax': syntax,
    }, self, GLOBALS.INVALIDATE_BUFFER_CACHE | GLOBALS.INVALIDATE_HIGHLIGHTS | GLOBALS.CHANGED)

  def finish(self):
    should_nc = self.should_nc

    if M.current == self and M.links > 0:
      self.linked = True

    if self.blocked or (M.links > 0 and M.diffs > 1 and self.linked):
      self.should_nc = should_nc = False
      self.blocked = True

    self.state |= _update_state({
      'nc': self.should_nc == True,
    }, self, GLOBALS.CHANGED)

    basebg = GLOBALS.basebg
    basebg_key = ''
    if callable(basebg):
      basebg = basebg(win, M.current)
    if basebg:
      basebg = COLOR_UTIL.to24b(basebg)
      basebg_key = 'bg(' + str(basebg) + ')'
    self.basebg = basebg if type(basebg) == int else None

    rerun_style = False
    style = self.style
    global_style = GLOBALS.style
    _global_style = self._global_style
    if len(_global_style) != len(global_style):
      rerun_style = True
    else:
      for i, s in enumerate(global_style):
        if _global_style[i] != s:
          rerun_style = True
          break
    if rerun_style == True:
      self.style = style = []
      self._global_style = _global_style =  []
      for s in global_style:
        _global_style.append(s)
        style.append(s.attach(self, self.style_state))

    hi_key = ''
    style_active = 0

    if self.blocked == False:
      for i, s in enumerate(style):
        s.before(self, self.style_state)
        style_key = s.key(self, self.style_state)
        if len(style_key) > 0:
          if style_active > 0:
            hi_key = hi_key + '#'
          hi_key = hi_key + style_key
          style_active += 1

    # TODO ensure this stays as is, signsretentionperiod is essentially based off win/buf change
    if self.nc == True:
      if GLOBALS.enablesigns and (GLOBALS.signsretentionperiod or 0) > 0 and (GLOBALS.now - self.timestamps['nc']) * 1000 < GLOBALS.signsretentionperiod:
        self.state |= GLOBALS.SIGNS

    # hi_key is used by the highlighter for replacement highlights.  These are
    # calculated based on a number of window criteria such as fadelevel, tint
    # wincolor.

    hi_key = '-'.join([str(x if x != None else 'N') for x in self.wincolorhl.values()]) + ':' + hi_key + ':' + basebg_key

    # INVALIDATE matches to be readded, free highlights, and get new synID
    self.state |= _update_state({
      'hi_key': hi_key,
      'matchpriority': 9 if self.is_minimap else int(GLOBALS.matchpriority),
    }, self, GLOBALS.INVALIDATE_HIGHLIGHTS | GLOBALS.CHANGED)

    # force the window to refresh if ncmode='windows' -- caching mostly handles performance here
    # We also hi linked status when nc_windows is activated.  This handles scenarios where the user
    # is on the diff side, against a buffer shared amongst multiple windows.
    self.style_active = style_active

    if (self.is_active_buf or (self.linked and GLOBALS.nc_windows)):
      self.state |= GLOBALS.CHANGED

    state = (self.state | GLOBALS.tick_state)
    if self.ns:
      if (GLOBALS.INVALIDATE_HIGHLIGHTS & state) > 0:
        self.ns.invalidate_highlights((GLOBALS.CHANGED & state) > 0)
      if (GLOBALS.INVALIDATE_BUFFER_CACHE & state) > 0:
        self.ns.invalidate_buffer_cache()
      if (GLOBALS.CHANGED & state) > 0:
        if self.style_active > 0:
          self.ns.highlight()
        else:
          self.ns.unhighlight()
      # elif here because we don't need to trigger SIGNS if CHANGED already occurred
      elif (GLOBALS.SIGNS & state) > 0:
        if self.style_active > 0:
          self.ns.add_signs()
        else:
          self.ns.remove_signs()
